### DISCUSSION TITLE
(not) Finding substructures

### DISCUSSION THREAD

**Discussion by NadavQuris** [rdkit/rdkit#6649] [Q&A] [answered]
I'm using rdkit to get substructure matches, but I'm not getting the results I expect to get. For example, take acetaminophen and the substructure hydroquinone:
```python
import rdkit

acetaminophen_inchi = "InChI=1S/C8H9NO2/c1-6(10)9-7-2-4-8(11)5-3-7/h2-5,11H,1H3,(H,9,10)"
aceta_mol = rdkit.Chem.MolFromInchi(acetaminophen_inchi)

hydroquinone_smarts = "[OH]c1ccc([OH,NH2,NH])cc1"
hydroquinone_mol = rdkit.Chem.MolFromSmarts(hydroquinone_smarts)
```

(Images of resulting molecule and substructure attached)
![image](https://github.com/rdkit/rdkit/assets/103879485/26ba22d2-c11d-48d7-a15a-2fdb718324e1)
![image](https://github.com/rdkit/rdkit/assets/103879485/5620682f-d375-4e06-a76e-c7e0fbb1138b)

Both the inchi and the smarts are taken from chembl. I would expect to see a match, but using: aceta_mol.GetSubstructMatches(hydroquinone_mol)  gives an empty result. Am I doing something wrong? Is this not a substructure?
I'm also attaching an image from chembl where they do identify this as a substructure:
![image](https://github.com/rdkit/rdkit/assets/103879485/eafa44dd-9287-4137-b3a7-70ae602bf9f3)

**Comment by ricrogz** [comment, marked as answer]
Hi @NadavQuris , the issue is that in your `hydroquinone_smarts` the N atoms are "capped" with hydrogens. It won't match the N in acetaminophen, which doesn't have any H atoms attached, having double bond instead of them. Removing these Hydrogens allows the matching:

```
import rdkit

acetaminophen_inchi = "InChI=1S/C8H9NO2/c1-6(10)9-7-2-4-8(11)5-3-7/h2-5,11H,1H3,(H,9,10)"
aceta_mol = rdkit.Chem.MolFromInchi(acetaminophen_inchi)

hydroquinone_smarts = ""[OH]c1ccc([OH,N])cc1""
hydroquinone_mol = rdkit.Chem.MolFromSmarts(hydroquinone_smarts)

aceta_mol.GetSubstructMatches(hydroquinone_mol) 
# ((10, 7, 3, 1, 6, 8, 2, 4),)
```

Note that the "matching" image from ChemBL does indeed have a H atom attached to its N atom.

**Reply by NadavQuris** [reply]
Thanks for the answer! This is something I'm doing with many molecules and many substructures, so I can't just change things manually. I tried using `Chem.RemoveHs` and `Chem.RemoveAllHs`, but that still resulted in no matches:

```python
import rdkit

acetaminophen_inchi = "InChI=1S/C8H9NO2/c1-6(10)9-7-2-4-8(11)5-3-7/h2-5,11H,1H3,(H,9,10)"
aceta_mol = rdkit.Chem.MolFromInchi(acetaminophen_inchi)

hydroquinone_smarts = "[OH]c1ccc([OH,NH2,NH])cc1"
hydroquinone_mol = rdkit.Chem.MolFromSmarts(hydroquinone_smarts)
hydroquinone_mol = rdkit.Chem.RemoveAllHs(hydroquinone_mol)

aceta_mol.GetSubstructMatches(hydroquinone_mol)

# ()
```

Any ideas?

**Reply by NadavQuris** [reply]
Nevermind, I answered myself. This did the job:

```python
def smarts_without_hydrogens(smarts_str):
    mol = Chem.MolFromSmarts(smarts_str)
    smiles_str = Chem.MolToSmiles(mol)
    return smiles_str
```

**Reply by ricrogz** [reply]
Please note that this is not a good idea at all. It is working for the case you suggested because SMILES doesn't support the "OR" query in the initial SMARTS (the `[OH,NH2,NH]` part), and RDKit replaces it with a dummy atom (`*`), which is interpreted as a wildcard when matching substructures.

This will, on one side, probably give you false positives (since you are no longer restricting the match to "O" or "N" atoms, but just any atom, including "C"), and it will only work for SMARTS patterns that include these "OR" queries, which it might not always be useful to replace with wildcards.

**Reply by NadavQuris** [reply]
I see. Do you have an alternative perhaps? I must find a solution to do this in a generic way.

**Reply by ricrogz** [reply]
I'm not really familiar with ChemBL or how you are selecting your SMARTS expressions, so I can't be very helpful.

Still, I'd recommend you to use SMILES instead of Inchi as the source of your structures, since, due to the way Inchi works, the resulting structures may be somewhat distorted. E.g., note that in the first image you posted, the N atom has a double bond attached, and no H, while the in the last image you posted, the one from ChemBL, the double bond is not attached to the N atom, but the O one, and the N atom has one H attached (so that it would have matched your SMARTS). SMILES is more consistent than Inchi, and the resulting structures should match the ones depicted in ChemBL.

This doesn't mean, though, that you won't have trouble with other structures. It all depends on how you define your SMARTS.
